### PR TITLE
Api key feature

### PR DESCRIPTION
**Pull request by tu55eladd**
This request features a basic user interface for registering a user with name, email and a domain for where the api is used. It also contains some queries for saving a user and key with it's owner ( a user ) and some endpoints for registering a user and checking if a given key is valid with hmac. For testing purposes the current hmac secret is not hidden.

### CODE REVIEW

File: server/knexfile.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    pool: {
+      min: 2,
+      max: 10
+    },
+    migrations: {
+      tableName: 'knex_migrations'
+    }
+  },
+
+  production: {
```
Comment (dagingaa) [inline server/knexfile.js]: Remove production and staging from this config, we aren't using it yet :)

File: server/dbconfig.js [hunk 2]
```diff
@@ -0,0 +1,5 @@
+var config = require('./knexfile.js');
+var env = 'development';
+var knex = require('knex')(config[env]);
+
+module.exports = knex;
```
Comment (dagingaa) [inline server/dbconfig.js]: I'm thinking we need to rename this file to something a bit better. How about knexStore.js?

File: server/repo/userRepo.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    .catch(function(error){
+        console.log(error);
+    });
+  };
+
+  static isValid(user){
+    if(!user || !user.name || !user.email || !user.domain){
+      return false;
+    }
+    console.log("Saving user");
```
Comment (dagingaa) [inline server/repo/userRepo.js]: We should get a proper logging framework for this, and not use console directly. Have a look at https://github.com/winstonjs/winston or something similar. If you don't need this log line, just delete it :)

File: server/repo/userRepo.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
+var knex = require('../dbconfig');
+
+class UserRepo {
+
+  static saveUser(user){
+    if(!this.isValid(user)){
+      console.log('rejecting because invalid user');
```
Comment (dagingaa) [inline server/repo/userRepo.js]: You don't need this log line, when you have the Promise below stating the error?

File: server/repo/userRepo.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
+var knex = require('../dbconfig');
+
+class UserRepo {
+
+  static saveUser(user){
```
Comment (dagingaa) [inline server/repo/userRepo.js]: Why do you use static here? I didn't think that was part of ES6 yet either? (I might be very wrong there)

File: server/repo/userRepo.js (hunk 3 above)
Comment (dagingaa) [inline server/repo/userRepo.js]: Besides, are you saving the user in this function anyway?

File: server/repo/userRepo.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+      domain: user.domain
+    })
+    .returning('id')
+    .catch(function(error){
+        console.log(error);
+    });
+  };
+
+  static isValid(user){
+    if(!user || !user.name || !user.email || !user.domain){
```
Comment (dagingaa) [inline server/repo/userRepo.js]: `return user && user.name && user.email && user.domain`

However, in other code we've written we ususally use `assert.ok(thingToCheck, "error message to throw")`. Would recommend checking that out! One for each test you're making makes the error messages you get back very accurate

File: server/migrations/20160617102450_keys.js [hunk 7]
```diff
@@ -0,0 +1,11 @@
+
+exports.up = function(knex, Promise) {
+  return knex.schema.createTable('keys', function(table){
+    table.string('key').primary();
+    table.integer('ownerId').references('id').inTable('users');
```
Comment (dagingaa) [inline server/migrations/20160617102450_keys.js]: I would like ownerId to be userId I think, easier to spot where it comes from I think

File: server/repo/keyRepo.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
+var knex = require('../dbconfig');
+
+class KeyRepo {
+
+  static saveKey(key, keyOwnerId){
+    if(!key){return null}
+
+    return knex('keys')
```
Comment (dagingaa) [inline server/repo/keyRepo.js]: Make "keys" a constant and use that

`const TABLE_NAME = "keys"`

File: server/repo/keyRepo.js [hunk 9]
```diff
@@ -0,0 +1,52 @@
+var knex = require('../dbconfig');
+
+class KeyRepo {
+
+  static saveKey(key, keyOwnerId){
+    if(!key){return null}
```
Comment (dagingaa) [inline server/repo/keyRepo.js]: Make this over two lines

**Comment by dagingaa** [comment]
I think we should introduce eslint before I continue reviewing this, that enforces some common standard on things and catches a lot of bugs I would miss otherwise. Basically copy the eslint setup from https://github.com/comoyo/video-conference/blob/master/.eslintrc and add a script in package.json for running eslint. `npm run lint`

File: server/keygen.js [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+    let keyWithHmac =
+    {
+        key: key,
+        hmac: this.generateHmac(key)
+    }
+    return JSON.stringify(keyWithHmac).toString('base64');
+  }
+
+  static guid() {
+    return this.s4() + this.s4() + '-' + this.s4() + '-' + this.s4() + '-' +
```
Comment (dagingaa) [inline server/keygen.js]: You could probably use https://www.npmjs.com/package/node-uuid instead, it's spec compliant and guaranteed unique

File: server/repo/userRepo.js (hunk 5 above)
Comment (tu55eladd) [inline server/repo/userRepo.js]: It is required when class is not instanciated ( does not work without it ).

Comment (dagingaa) [inline server/repo/userRepo.js]: I would make it instantiatable, and pass in the knexStore in the constructor. See how we write database classes in https://github.com/comoyo/video-conference/blob/master/modules/stores/deviceStore.js

File: app/RegisterComponent.jsx [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    event.preventDefault();
+    let userData = this.getFormValues(event.target);
+
+    $.ajax({
+      type: 'POST',
+      url: '/register',
+      contentType: 'application/json; charset=utf-8',
+      data: JSON.stringify(userData),
+      success: (res, textStatus, xhr)=>{
+        console.log(res);
```
Comment (dagingaa) [inline app/RegisterComponent.jsx]: Either make this log message more explicit, or remove it

File: app/RegisterComponent.jsx [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+    super(props);
+    autoBind(this);
+    this.state = { userData: {} };
+  }
+
+  onRegister(event){
+    event.preventDefault();
+    let userData = this.getFormValues(event.target);
+
+    $.ajax({
```
Comment (dagingaa) [inline app/RegisterComponent.jsx]: Don't use jquery anymore, use the new fetch API. Here is a polyfill: `npm install whatwg-fetch --save`

File: app/RegisterComponent.jsx [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+      name: targetForm.children[0].value,
+      email: targetForm.children[3].value,
+      domain: targetForm.children[6].value
+    }
+  }
+
+  render(){
+    return(
+        <form onSubmit={this.onRegister} >
+          <input type="text" placeholder="Name"/><br/><br/>
```
Comment (dagingaa) [inline app/RegisterComponent.jsx]: Don't use line breaks here, either fix it with CSS or keep it ugly and remove them regardless

File: app/RegisterComponent.jsx [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+      type: 'POST',
+      url: '/register',
+      contentType: 'application/json; charset=utf-8',
+      data: JSON.stringify(userData),
+      success: (res, textStatus, xhr)=>{
+        console.log(res);
+        this.props.onSubmit(res);
+      },
+      error: (xhr, status, err)=>{
+        console.log(err);
```
Comment (dagingaa) [inline app/RegisterComponent.jsx]: Same here, make the log message more explicit, or remove

File: app/KeyComponent.jsx [hunk 15]
```diff
@@ -0,0 +1,20 @@
+import React from 'react';
+import autoBind from 'react-autobind';
+
+class KeyComponent extends React.Component {
+
+  constructor(props){
+    super(props);
+    autoBind(this);
+    console.log(props);
```
Comment (dagingaa) [inline app/KeyComponent.jsx]: Do you need to console.log here?

File: app/KeyComponent.jsx [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+  }
+
+  render(){
+    return(
+      <div>Your key : {this.props.apiKey}</div>
```
Comment (dagingaa) [inline app/KeyComponent.jsx]: Nitpick: remove the space before `:`

File: app/ProcessViewComponent.jsx [hunk 17]
```diff
@@ -0,0 +1,36 @@
+import React from 'react';
+import autoBind from 'react-autobind';
+
+class ProcessViewComponent extends React.Component{
```
Comment (dagingaa) [inline app/ProcessViewComponent.jsx]: Did you end up using this class?

File: app/ProcessViewComponent.jsx [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+      step : props.data[0],
+      stepIndex: 0
+    }
+  }
+
+  render(){
+    let stepNodes = this.props.data.map( (step, index)=> {
+      if(index==this.props.stepIndex){
+        return(
+        <li style={{display: 'inline-block', paddingRight: '30px'}} ><h3>{step}</h3></li>
```
Comment (dagingaa) [inline app/ProcessViewComponent.jsx]: Don't do inline styles like this, either fix it properly with CSS, or use some of the new-fancy jsx CSS definitions. But don't code it straight into the template like this

File: app/ProcessViewComponent.jsx [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+  constructor(props){
+    super(props);
+    this.state = {
+      step : props.data[0],
+      stepIndex: 0
+    }
+  }
+
+  render(){
+    let stepNodes = this.props.data.map( (step, index)=> {
```
Comment (dagingaa) [inline app/ProcessViewComponent.jsx]: Formatting here is off. Write `this.props.data.map((step, index) => {`

File: app/ProcessViewComponent.jsx [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+      if(index==this.props.stepIndex){
```
Comment (dagingaa) [inline app/ProcessViewComponent.jsx]: This needs some breathing room, `if (index === this.props.stepIndex) {`. In addition, always use `===`

File: app/ProcessViewComponent.jsx [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+        return(
+        <li style={{display: 'inline-block', paddingRight: '30px'}} ><h3>{step}</h3></li>
+        )
+      }
+      return(
+        <li style={{display: 'inline-block', paddingRight: '30px', color: 'grey'}} ><h3>{step}</h3></li>
```
Comment (dagingaa) [inline app/ProcessViewComponent.jsx]: Remove inline styles as described above

File: app/index.jsx [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+    this.getKeyTabName = "Get key";
+    this.registerTabText = "Register";
+    this.getKeyTabText = "Get key";
+
+
+    this.state = {
+      stepNames: ['Register', 'Get your key'],
+      selectedTab:
+        <RegisterComponent
+        setKey={this.setKeyValue}
```
Comment (dagingaa) [inline app/index.jsx]: Indent these a bit more, makes it clearer that they are part of the HTML tag above

File: app/index.jsx [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+      }
+
+  }
+
+  onSubmit(keyVal){
+    this.setKeyTab(keyVal);
+  }
+
+  setRegisterTab(){
+    this.setState({selectedTab:
```
Comment (dagingaa) [inline app/index.jsx]: Do you need to duplicate this HTML element? Stick to DRY, and see if you can't pull it out so it isn't duplicated

File: app/index.jsx [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+      <RegisterComponent
+      setKey={this.setKeyValue}
+      onSubmit={this.onSubmit}/>
+    });
+  }
+
+  setKeyTab(keyVal){
+    console.log("Setting key tab");
```
Comment (dagingaa) [inline app/index.jsx]: Do you need these log lines?

File: app/index.jsx [hunk 25]
```diff
@@ -0,0 +1,62 @@
...
+    console.log(keyVal);
+    this.setState({selectedTab:
+      <KeyComponent apiKey={keyVal} />,
```
Comment (dagingaa) [inline app/index.jsx]: Why is onSubmit removed here?

File: app/index.jsx [hunk 26]
```diff
@@ -10,11 +10,6 @@ class App extends React.Component {
   constructor(props){
     super(props);
     autoBind(this);
-    this.registerTabName = "Register";
```
Comment (dagingaa) [inline app/index.jsx]: If this is removed, then merge it together with the original commit

File: app/RegisterComponent.jsx [hunk 27]
```diff
@@ -1,5 +1,6 @@
 import React from 'react';
 import autoBind from 'react-autobind';
+import $ from 'jquery';
```
Comment (dagingaa) [inline app/RegisterComponent.jsx]: Don't use jquery. Also, if this is required in another commit, then add it to the commit that needs it. Try to make all the commits runnable by themselves, where you can

File: server/server.js [hunk 28]
```diff
@@ -34,10 +34,6 @@ app.post('/register', jsonParser, (req, res)=>{
   });
 })
 
-app.post('/login', (req, res) => {
```
Comment (dagingaa) [inline server/server.js]: merge into original commit

File: templates/index.html [hunk 29]
```diff
@@ -56,7 +57,10 @@ <h1 class="main-title outline">appear.in</h1>
                 <header class="content-header">
                     <h1 class="title">Embed appear.in on your web page</h1>
                 </header>
-                {{{ contents }}}
+                  {{{ contents }}}
+            </section>
+            <section class="content">
+              <div id="app" />
```
Comment (dagingaa) [inline templates/index.html]: make sure this commit comes before the document.getElementById("app") you have in a previous commit, or merge it into that commit.

File: package.json [hunk 30]
```diff
@@ -33,7 +33,7 @@
     "react-autobind": "1.0.6",
 
     "knex": "^0.11.5",
-    "sqlite3": "^3.1.4",
+    "sqlite3": "^3.1.4"
```
Comment (dagingaa) [inline package.json]: Merge into previous commit

File: server/keygen.js [hunk 31]
```diff
@@ -31,4 +34,17 @@ class KeyGenerator {
 
 }
 
+/* Copied and pasted from mozilla developer pages as a solution for the "Unicode problem" */
+function b64EncodeUnicode(str) {
```
Comment (dagingaa) [inline server/keygen.js]: Can you link to the page where thi was found? Also, do we encode unicode anywhere? I don't think this is needed, but I would like to learn more why it's needed.

File: server/keygen.js [hunk 32]
```diff
@@ -12,7 +12,7 @@ class KeyGenerator {
 
     let keyString = b64EncodeUnicode(JSON.stringify(keyWithHmac));
 
-    return keyString);
+    return keyString;
```
Comment (dagingaa) [inline server/keygen.js]: merge into commit where it was introduced

File: server/keygen.js [hunk 33]
```diff
@@ -34,17 +34,4 @@ class KeyGenerator {
 
 }
 
-/* Copied and pasted from mozilla developer pages as a solution for the "Unicode problem" */
```
Comment (dagingaa) [inline server/keygen.js]: merge into commit where it was introduced. perhaps you can just remove that commit now?

File: server/server.js [hunk 34]
```diff
@@ -35,16 +37,27 @@ app.post('/register', jsonParser, (req, res)=>{
 })
 
 app.post('/isvalidkey', jsonParser, (req, res)=>{
-  let key = req.body.key;
-  let hmac = req.body.hmac;
+  let body = Buffer.from(req.body.key, 'base64').toString('ascii');
+  try{
+      let data = JSON.parse(body)
```
Comment (dagingaa) [inline server/server.js]: I don't think you need this, bodyparser should do this automatically.

File: app/KeyComponent.jsx [hunk 35]
```diff
@@ -0,0 +1,19 @@
...
+class KeyComponent extends React.Component {
+
+  constructor(props){
+    super(props);
+    autoBind(this);
+  }
+
+  render(){
+    return(
+      <div className="key" >Your key :{this.props.apiKey}</div>
```
Comment (dagingaa) [inline app/KeyComponent.jsx]: This should probably read "Your key: {this.props.apiKey}", otherwise a user might think the `:` is part of their key

File: app/RegisterComponent.jsx [hunk 36]
```diff
@@ -0,0 +1,53 @@
...
+  constructor(props){
+    super(props);
+    autoBind(this);
+    this.state = { userData: {} };
+  }
+
+  onRegister(event){
+    event.preventDefault();
+
+    let userData = JSON.stringify(this.getFormValues(event.target));
```
Comment (dagingaa) [inline app/RegisterComponent.jsx]: This can be sent as raw JSON, and will avoid you having to manually parse it in the backend. So avoid doing JSON.stringify here, that will already be done for you

Comment (dagingaa) [inline app/RegisterComponent.jsx]: Actually, I might be wrong. That's a bit unexpected o0: https://github.com/github/fetch#post-json

File: app/RegisterComponent.jsx [hunk 37]
```diff
@@ -0,0 +1,53 @@
...
+
+    fetch('/register', {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: userData
+    })
+    .then((res)=> {
+      return res.text();
```
Comment (dagingaa) [inline app/RegisterComponent.jsx]: You can optionally keep this as a single line, like this:
`.then(res => res.text())`

File: app/RegisterComponent.jsx [hunk 38]
```diff
@@ -0,0 +1,53 @@
...
+    })
+    .then((key)=>{
+      this.props.onSubmit(key);
```
Comment (dagingaa) [inline app/RegisterComponent.jsx]: It's a bit hard to follow the flow here, is this the apiKey? Why is that sent to something called onSubmit? Maybe it should be onSuccess or something similar?

File: files/assets/_register.scss [hunk 39]
```diff
@@ -0,0 +1,14 @@
+.process-step{
```
Comment (dagingaa) [inline files/assets/_register.scss]: The CSS here is a bit generic, perhaps we should namespace them a bit to avoid conflicts later?

for example:
```
.registration-form {
    .process-step { .. }
    .unactive-step { .. }
}
```

File: package.json [hunk 40]
```diff
@@ -4,18 +4,49 @@
...
+    "node-uuid": "^1.4.7",
+    "react": "^15.1.0",
+    "react-autobind": "1.0.6",
+    "react-dom": "^15.1.0",
+    "serve-static": "^1.11.1",
+    "sqlite3": "^3.1.4",
+    "winston": "^2.2.0",
+    "knex": "^0.11.5",
+    "sqlite3": "^3.1.4",
+
```
Comment (dagingaa) [inline package.json]: Nitpick, remove the spaces here

File: server/keygen.js [hunk 41]
```diff
@@ -0,0 +1,32 @@
...
+        key: key,
+        hmac: this.generateHmac(key)
+    };
+
+    let keyData = new Buffer(JSON.stringify(keyWithHmac)).toString('base64');
+
+    return keyData;
+  }
+
+  static guid() {
```
Comment (dagingaa) [inline server/keygen.js]: This isn't a guid anymore :)

File: server/keygen.js [hunk 42]
```diff
@@ -0,0 +1,32 @@
+const crypto = require('crypto');
+var uuid = require('node-uuid');
+
+class KeyGenerator {
+
+  static generate(){
+    let key = this.uuid();
```
Comment (dagingaa) [inline server/keygen.js]: This function is undefined

File: server/keygen.js (hunk 41 above)
Comment (dagingaa) [inline server/keygen.js]: looks like you want to rename this into `getApiId` or something similar. Don't name it after what it returns, like a uuid, but after what it is, for example an apiId or whatever you choose to call it

File: server/repo/keyRepo.js [hunk 43]
```diff
@@ -0,0 +1,59 @@
...
+
+class KeyRepo {
+
+  constructor(knex){
+    this.knex = knex;
+  }
+
+  saveKey(key, keyOwnerId){
+    if(!key){
+      return null;
```
Comment (dagingaa) [inline server/repo/keyRepo.js]: Is this a programming error? In which case you probably want to throw. Use assert.ok for example to handle input validation

File: server/repo/keyRepo.js [hunk 44]
```diff
@@ -0,0 +1,59 @@
...
+      log.log('error', error);
+    });
+  }
+
+  getKey(userId){
+    return this.knex
+    .select('key')
+    .from('keys')
+    .where('ownerId', '=', userId)
+    .catch( (error)=>{
```
Comment (dagingaa) [inline server/repo/keyRepo.js]: nitpick, no need for a space here

File: server/repo/keyRepo.js [hunk 45]
```diff
@@ -0,0 +1,59 @@
...
+    if(!key){
+      return null;
+    };
+
+    return this.knex(TABLE_NAME)
+    .insert({
+      key: key,
+      ownerId: keyOwnerId
+    })
+    .catch((error)=>{
```
Comment (dagingaa) [inline server/repo/keyRepo.js]: but let the arrow breathe a bit here

File: server/server.js [hunk 46]
```diff
@@ -0,0 +1,59 @@
...
+var log = require('winston');
+var app = express();
+
+
+var jsonParser = bodyParser.json();
+
+app.use(express.static("public"));
+app.use(express.static("build"));
+app.use(bodyParser.json({ type: 'application/*+json' }));
+app.use(bodyParser.urlencoded({extended: true}));
```
Comment (dagingaa) [inline server/server.js]: Hmm, I haven't seen these two configuration options before, what do they do?

File: server/server.js [hunk 47]
```diff
@@ -0,0 +1,59 @@
+var UserRepo = require('./repo/userRepo.js');
+var KeyRepo = require('./repo/keyRepo.js');
+var KeyGenerator = require('./keygen.js');
+var express = require('express');
+var bodyParser = require('body-parser');
+var log = require('winston');
+var app = express();
+
+
+var jsonParser = bodyParser.json();
```
Comment (dagingaa) [inline server/server.js]: Do you still need this when you have added bodyParser as a middleware in line 14?

File: server/server.js [hunk 48]
```diff
@@ -0,0 +1,59 @@
...
+    .then(()=>{
+        res.send(key);
+    })
+    .catch((error)=>{
+      log.log('error', error);
+    });
+  });
+});
+
+app.post('/isvalidkey', jsonParser, (req, res)=>{
```
Comment (dagingaa) [inline server/server.js]: we should probably make this a bit mor restful, like `/key/validate`

File: server/server.js [hunk 49]
```diff
@@ -0,0 +1,59 @@
...
+  let body = Buffer.from(req.body.key, 'base64').toString('ascii');
+  try{
+      let data = JSON.parse(body);
```
Comment (dagingaa) [inline server/server.js]: The naming here is a bit weird, I think we should clean up this function. Why do you have req.body.key stored to a body variable, and then that is JSON parsed into data? Improving the variable naming eases readability. How about `decodedApiCredentials` and `credentials` or something? We should figure out what we want to name each of these so they make more sense. Naming is hard, but important!

File: app/RegisterComponent.jsx (hunk 37 above)
Comment (tu55eladd) [inline app/RegisterComponent.jsx]: Maybe it is more consistent keeping it considering the next "then-clause" ?

File: server/server.js [hunk 50]
```diff
@@ -0,0 +1,59 @@
...
+  let key = KeyGenerator.generate();
+
+  UserRepo.saveUser(req.body)
+  .then((ownerId)=>{
+    KeyRepo.saveKey(key, ownerId)
+    .then(()=>{
+        res.send(key);
+    })
+    .catch((error)=>{
+      log.log('error', error);
```
Comment (dagingaa) [inline server/server.js]: The error messages here should be a bit more verbose I think. What actually fails here? In this case, it seems it will fail if you can't save the key. Then the error message should probably read: "Error: Cannot save key"  and then print the actual error

File: server/server.js [hunk 51]
```diff
@@ -0,0 +1,59 @@
...
+app.get('/', (req, res) => {
+  res.render('index');
+});
+
+app.post('/register', jsonParser, (req, res)=>{
+  let key = KeyGenerator.generate();
+
+  UserRepo.saveUser(req.body)
+  .then((ownerId)=>{
+    KeyRepo.saveKey(key, ownerId)
```
Comment (dagingaa) [inline server/server.js]: I have a hard time following what happens here. If you do not return here, the promise chain is broken, but it seems express doesn't really care as long as you don't send `res.send()`. However, if the `.catch` is triggered, you don't send anything back, making the request hang until it times out. Is this intended?

File: files/assets/_register.scss (hunk 39 above)
Comment (tu55eladd) [inline files/assets/_register.scss]: They are not part of the registration form, but i will find another name

File: app/RegisterComponent.jsx (hunk 37 above)
Comment (dagingaa) [inline app/RegisterComponent.jsx]: I agree, let's keep it this way for now

File: server/server.js [hunk 52]
```diff
@@ -0,0 +1,76 @@
...
+      log.log('error', error);
+    });
+
+  });
+});
+
+app.post('/key/validate', jsonParser, (req, res)=>{
+  let decodedApiCredentials = Buffer.from(req.body.apiKey, 'base64').toString('ascii');
+  try{
+      let credentials = decodedApiCredentials;
```
Comment (dagingaa) [inline server/server.js]: are you sure this still works? You removed the JSON.parse here, which I think you need...

**Comment by dagingaa** [comment]
There's still more we can clean up here, but we can fix those things later to be able to move forward with this.

Good job so far! 💃 

🍰
